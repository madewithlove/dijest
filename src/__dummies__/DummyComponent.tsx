import * as React from 'react';

const DummyComponent: React.SFC<{ text?: string }> = ({ text }) =>
    <div>
        <div className="some-div">hello</div>
        <button>
            {text || 'foobar'}
        </button>
    </div>;

export default DummyComponent;
