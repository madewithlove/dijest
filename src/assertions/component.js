import renderer from 'react-test-renderer';

/**
 * Assert that a component equals X
 */
export default actual => {
    if (typeof actual.toJSON === 'undefined') {
        actual = renderer.create(actual);
    }

    const tree = actual.toJSON();
    expect(tree).toMatchSnapshot();

    return {actual, tree};
};
