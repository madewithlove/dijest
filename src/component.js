import {mount, render} from 'enzyme';
import {matchesSnapshot} from './helpers/matchesSnapshot';
import toEnzymeWrapper from './helpers/toEnzymeWrapper';

export default function component(description, tested, callback) {
    if (!tested) {
        return matchesSnapshot(toEnzymeWrapper(description));
    }

    it(description, () => {
        const wrapper = toEnzymeWrapper(tested, callback ? mount : render);
        if (callback) {
            callback(wrapper, matchesSnapshot);
        } else {
            matchesSnapshot(wrapper);
        }
    });
}
