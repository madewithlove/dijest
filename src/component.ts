import { mount, render } from 'enzyme';
import matchesSnapshot from './helpers/matchesSnapshot';
import toEnzymeWrapper, { EnzymeWrapper } from './helpers/toEnzymeWrapper';
import { ComponentClass, ReactElement, StatelessComponent } from 'react';
import withJestOptions, { JestOptions } from './helpers/withJestOptions';
import ComponentTester from './Testers/ComponentTester';

export type ComponentFactory =
    | ReactElement<any>
    | StatelessComponent
    | ComponentClass;

export type ComponentCallback = (
    wrapper: EnzymeWrapper,
    matchesSnapshot: Function,
) => void;

function component(
    description: string | ComponentFactory,
    tested?: ComponentFactory,
    callback?: ComponentCallback | string,
    it?: jest.It,
) {
    const tester = new ComponentTester().setTester(it);
    const numberArgs = [description, tested, callback].filter(
        value => typeof value !== 'undefined',
    ).length;

    switch (true) {
        case numberArgs === 1 && typeof description !== 'string':
            return tester.setComponent(description as ComponentFactory).run();

        case numberArgs === 2:
            return tester
                .setDescription(description as string)
                .setComponent(tested)
                .run();

        case numberArgs === 3 && typeof callback === 'string':
            return tester
                .setDescription(description as string)
                .setComponent(tested)
                .onlySelector(callback as string)
                .run();

        case numberArgs === 3 && typeof callback === 'function':
            return tester
                .setDescription(description as string)
                .setComponent(tested)
                .setCallback(callback as Function)
                .run();
    }
}

export default withJestOptions(component);
