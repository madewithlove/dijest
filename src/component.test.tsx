import * as React from 'react';
import component from './component';
import DummyComponent from './__dummies__/DummyComponent';

component('can create snapshots easily', <DummyComponent text="bar" />);

component("can accept a component factory if you're lazy", DummyComponent);

component(
    'can snapshot part of component',
    <DummyComponent text="bar" />,
    '.some-div',
);

component(
    'provides access to the Enzyme wrapper if need be',
    <DummyComponent text="bar" />,
    (wrapper, snapshot) => {
        // You can snapshot the wrapper at any type
        snapshot(wrapper.find('.some-div'));

        // Instance of Enzyme's mount wrapper
        wrapper.find('button').simulate('click');
        snapshot(wrapper);
    },
);
