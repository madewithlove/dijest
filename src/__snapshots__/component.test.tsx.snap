// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can pass component as function 1`] = `
<div>
  <p>
    foobar
  </p>
</div>
`;

exports[`can receive wrapper 1`] = `
<p>
  foobar
</p>
`;

exports[`can write snapshot test 1`] = `
<div>
  <p>
    hello
  </p>
</div>
`;
